Add support for non aligned source address

sslupsky opened the pull request:
The PR addresses an issue with the sam0 flash driver.  The flash driver writes data to the flash page using 32 bit long words.  This will generate a hard fault if the source address for a flash write is not aligned to a 32 bit word boundary.  This PR fixes this issue.

sslupsky committed:
drivers: flash: flash_sam0:  Add support for non aligned source address

If a source address is not aligned with a 32bit data type
the existing code causes a hard fault.  This commit adds a
inline function to return the value from a non aligned source address.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

jfischer-phytec-iot commented:
@sslupsky please take a look at [project conding-style ](https://docs.zephyrproject.org/latest/contribute/index.html?highlight=coding#coding-style)

sslupsky committed:
drivers: flash: flash_sam0:  Update code format

Update code to conform with contrbution guidelines.

Signed-off-by: Steven Slupsky <sslupsky@gmail.com>

jukkar commented on drivers/flash/flash_sam0.c:
```diff
@@ -112,6 +112,23 @@ static int flash_sam0_check_status(off_t offset)
 	return 0;
 }
 
+static inline u32_t flash_sam0_read_unaligned_u32(const void *data)
```
Instead of this, can't you just use `UNALIGNED_GET()` macro?

nvlsianpu replied:
I was not sure if it does the same, but if yes - that better idea.

benpicco commented:
You don't need to do that for every byte.
Just the first (up top 3) bytes need to be copied manually, then you can continue copying whole words.

sslupsky commented:
@jukkar I looked into the UNALIGNED_GET() macro and that looks good ... thanks for pointing that out to me.  I'll do some testing.

@benpicco The source address could potentially be unaligned so all the memory reads need to account for this.  The destination address must always be aligned and be a 32 bit write (you cannot write a byte at a time).  The UNALIGNED_GET() macro @jukkar pointed out looks like it will eliminate the need for the flash_sam0_read_unaligned_u32() function.

sslupsky committed:
drivers:i2c: fixes incorrect declaration

This commit fixes an incorrect declaration of the
buffer_size member of the i2c_eeprom_slave_config
struct.

sslupsky commented:
Resubmitted a new PR #23114 using the UNALIGNED_GET() macro.
